deno_typescript/lib.rs

```diff
@@ -0,0 +1,322 @@
+
+  let config_json = serde_json::json!({
+    "compilerOptions": {
+      "declaration": true,
+      "lib": ["esnext"],
+      "module": "esnext",
+      "target": "esnext",
+      "listFiles": true,
+      "listEmittedFiles": true,
+      // Emit the source alongside the sourcemaps within a single file;
+      // requires --inlineSourceMap or --sourceMap to be set.
+      // "inlineSources": true,
+      "sourceMap": true,
+    },
+  });
+
+  let mut root_names_str: Vec<String> = root_names
+    .iter()
+    .map(|p| {
+      assert!(p.exists());
+      p.to_string_lossy().to_string()
+    })
+    .collect();
+  root_names_str.push("$asset$/lib.deno_core.d.ts".to_string());
+
+  // TODO lift js_check to caller?
+  let state = js_check(ts_isolate.compile(&config_json, root_names_str));
+
+  Ok(state)
+}
+
+pub fn compile_bundle(
+  bundle: &Path,
+  root_names: Vec<PathBuf>,
+) -> Result<Arc<Mutex<TSState>>, ErrBox> {
+  let ts_isolate = TSIsolate::new(true);
+
+  let config_json = serde_json::json!({
+    "compilerOptions": {
+      "declaration": true,
+      "lib": ["esnext"],
+      "module": "amd",
+      "target": "esnext",
+      "listFiles": true,
+      "listEmittedFiles": true,
+      // "types" : ["typescript.d.ts"],
+      "typeRoots" : ["$typeRoots$"],
+      // Emit the source alongside the sourcemaps within a single file;
+      // requires --inlineSourceMap or --sourceMap to be set.
+      // "inlineSources": true,
+      "sourceMap": true,
+      "outFile": bundle,
+    },
+  });
+
+  let mut root_names_str: Vec<String> = root_names
+    .iter()
+    .map(|p| {
+      if !p.exists() {
+        panic!("File not found {}", p.display());
+      }
+
+      let module_specifier =
+        ModuleSpecifier::resolve_url_or_path(&p.to_string_lossy()).unwrap();
+      module_specifier.as_str().to_string()
+      // p.to_string_lossy().to_string()
+    })
+    .collect();
+  root_names_str.push("$asset$/lib.deno_core.d.ts".to_string());
+
+  // TODO lift js_check to caller?
+  let state = js_check(ts_isolate.compile(&config_json, root_names_str));
+
+  Ok(state)
+}
+
+#[allow(dead_code)]
+fn print_source_code(code: &str) {
+  let mut i = 1;
+  for line in code.lines() {
+    println!("{:3}  {}", i, line);
+    i += 1;
+  }
+}
+
+/// Create a snapshot and store the resulting path in the provided enviromental
+/// variable.
+pub fn mksnapshot_bundle(
+  bundle: &Path,
+  env_var: &str,
+  state: Arc<Mutex<TSState>>,
+) -> Result<(), ErrBox> {
+  let mut runtime_isolate = Isolate::new(StartupData::None, true);
+  let source_code_vec = std::fs::read(bundle)?;
+  let source_code = std::str::from_utf8(&source_code_vec)?;
+
+  js_check(
+    runtime_isolate.execute("amd_runtime.js", include_str!("amd_runtime.js")),
+  );
+
+  js_check(runtime_isolate.execute(&bundle.to_string_lossy(), &source_code));
+
+  // execute main.
+
+  let state = state.lock().unwrap();
+  let main = state.written_files.last().unwrap().module_name.clone();
```

Comment:
Can this be abstracted as method on `TsState` with short comment? It's not obvious that last written file is first root name.

Reply:
Yes this is quite hacky and needs a comment. Fixed.